### ISSUE TITLE
About Deterministic Behaviour of GPU implementation of tensorflow

### ISSUE THREAD

**Issue by antares1987** [tensorflow/tensorflow#12871]
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Debian
- **TensorFlow version (use command below)**:('v1.3.0-rc2-20-g0787eee', '1.3.0')
- **CUDA/cuDNN version**: 8.0
- **GPU model and memory**: GeForce GTX 950

Dear experts,

I have just written a python code that implements a CNN with tensorflow. Despite using tf.set_random_seed(0) throughout my code, I get different results for each different run using a GPU, while this is not happening when switching to CPU. I read different threads about stochastic behaviour of GPUs and I could find that the stochastic behaviour is happening when using AdamOptimizer. My code now uses a CPU just for the optimizer, and a GPU for all the others operations, so now the time performance of the GPU is 25% better than CPU despite the 700% of the full GPU implementation. My question is: is this avoidable? Should I really give up in having a deterministic code when using GPUs? How can I tune my hyperparameters and use a full GPU code (some forums say to make an average of different runs.. but this kills the time advantage of using GPUs).

Thanks in advance for all your kind support and for the wonderful job

**Comment by drpngx** [comment]
There was some discussion previously #10636. Does that help?

**Comment by antares1987** [comment]
Hello, thank you for your reply. yes, indeed the document is useful, but I see that in the example there's no a minimization of the loss through the metod tf.minimize. Indeed I fount that the stochastic component of my code is in optimizer = tf.train.AdamOptimizer(rate).minimize(loss). Now, my question is, is there a way to use an optimizer and to get a deterministic result at the same time?

Thank you very much again

**Comment by reedwm** [comment]
@ekelsen do you know when determinism will be available?

**Comment by ekelsen** [comment]
Reductions are already deterministic as well as l2loss.  softmax will be soon,  Cross entropy following.

**Comment by ekelsen** [comment]
The gradient calculation of convolutions is not likely to be deterministic at anywhere close to the same performance unless NVIDIA decides to change the implementation of cuDNN.

**Comment by OutSorcerer** [comment]
@ekelsen, as far as I understand, Theano also uses cuDNN, but they still have settings (in .theanorc):
```
[dnn.conv]

algo_bwd_data=deterministic
algo_bwd_filter=deterministic
```
that [work with convolutions in a slower but deterministic way](http://deeplearning.net/software/theano/library/sandbox/cuda/dnn.html).

**Comment by ekelsen** [comment]
That's entirely correct and what I was referring to.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by drpngx** [comment]
Closing since it sounds like @ekelsen solved the problem.

**Comment by OutSorcerer** [comment]
@drpngx, the original issue with non-deterministic behaviour of TensorFlow still persists.

The root cause is in cuDNN, not Tensorflow, but it is still should be possible to make the behavior deterministic in the same way like Theano developers did with

```
[dnn.conv]

algo_bwd_data=deterministic
algo_bwd_filter=deterministic
```

@ekelsen said:

>The gradient calculation of convolutions is not likely to be deterministic at anywhere close to the same performance unless NVIDIA decides to change the implementation of cuDNN.

But in fact performance *is* almost the same.

According to my own experiments on Geforce gtx 1050 the performance is almost the same both for Tensorflow using cuDNN vs. Theano using cuDNN with deteministic/non-deterministic config (about 4s/epoch all all three cases).  

Other people also report relatively the same performance with Theano/cuDNN and with or without `algo_bwd_data` and `algo_bwd_filter` configuration:

* https://github.com/keras-team/keras/issues/2479#issuecomment-213987747
* https://www.kaggle.com/c/ultrasound-nerve-segmentation/discussion/22014

So I believe it still makes sense to implement something like that in TensorFlow.

**Comment by protoget** [comment]
You can try set TF_CUDNN_USE_AUTOTUNE = '0' and try again. Though prepare to experience degenerated performance. Determinism + High performance isn't what we'd support.

/cc @yzhwang

**Comment by lucb** [comment]
TF_CUDNN_USE_AUTOTUNE='0' will not work, it will use cuDNN function to retrieve the best algorithm. One solution I found was to support a new environment variable TF_CUDNN_CONVOLUTION_BWD_FILTER_ALGO_DETERMINISTIC='1' such that we only use cuDNN deterministic algorithms.

I did not measure the performance of this solution but I'm getting deterministic results for gradient computation of convolutions, something that I was not getting previously

The code I used to test the determinism of the backward filter of the convolution is the following:
```python
import tensorflow as tf
import numpy as np

np.random.seed(0)
np.set_printoptions(precision=4)

def run_convolution_test(device, N=1):
  tf.reset_default_graph()
  tf.set_random_seed(42)
  with tf.device(device):
    input = tf.random_normal([1, 2048, 2048, 1], name='input')
    kernel = tf.get_variable('kernel', shape = [3,3,1,1])
    conv = tf.nn.conv2d(name='conv', input=input,
                        filter=kernel, strides=[1,1,1,1], padding='SAME',
                        use_cudnn_on_gpu=True)
    grad = tf.gradients(conv, kernel)
  grad_val = None
  a_val = None
  config = tf.ConfigProto(intra_op_parallelism_threads=1, inter_op_parallelism_threads=1)
  with tf.Session(config=config) as sess:
    sess.run(tf.global_variables_initializer())
    conv_val, grad_val = sess.run([tf.squeeze(conv), tf.squeeze(grad)])
    return conv_val, grad_val

def compute_error_matrix(matrices):
    total_absolute_error = np.zeros((len(matrices), len(matrices)))
    for i in range(len(matrices)):
        for j in range(len(matrices)):
            if i != j:
                dif=matrices[i] - matrices[j]
                total_absolute_error[i][j] = np.sum(np.absolute(dif))
    print(total_absolute_error)

print("Running on GPU")
conv_gpu, grad_gpu = zip(*[run_convolution_test('/device:GPU:0', i) for i in range(10)])
compute_error_matrix(conv_gpu)
compute_error_matrix(grad_gpu)

print("Running on CPU")
conv_cpu, grad_cpu = zip(*[run_convolution_test('/device:CPU:0', i) for i in range(10,20)])
compute_error_matrix(conv_cpu)
compute_error_matrix(grad_cpu)
```

**Comment by yoavz** [comment]
+1 to @lucb - the optional enabling of deterministic cudNN algo's would be very useful. Even if a performance trade-off exists for determinism, users should have the ability to enable it.

**Comment by Lorraine333** [comment]
But it seems what the real problem is Adam optimizer is not deterministic. Is there any possible flag to make it deterministic for Adam?

Thanks so much.

**Comment by chithangduong** [comment]
I second this. My computation graph does not have any convolution operation but it still suffers from non-determinism. By sending the compute_gradients operation of AdamOptimizer to the CPU, the problem disappears.

**Comment by drpngx** [comment]
@chithangduong What kind of differences do you see? Could they be due to multi-threading?

**Comment by chithangduong** [comment]
@drpngx I have figured it out. The non-deterministic part is the [clip_by_value ](https://www.tensorflow.org/api_docs/python/tf/clip_by_value) operation. By sending it to CPU and setting colocate_gradients_with_ops for the AdamOptimizer, the results between two different runs are the same. Do you have any idea why this is the case?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @drpngx: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by drpngx** [comment]
Could you clarify which one you're using? Unary, binary (left or right) or ternary? Thanks.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by tensorflowbutler** [comment]
It has been 29 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by tensorflowbutler** [comment]
It has been 44 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by tensorflowbutler** [comment]
We are closing this issue for now due to lack of activity. Please comment if this is still an issue for you. Thanks!

**Comment by leriomaggio** [comment]
> TF_CUDNN_USE_AUTOTUNE='0' will not work, it will use cuDNN function to retrieve the best algorithm. One solution I found was to support a new environment variable TF_CUDNN_CONVOLUTION_BWD_FILTER_ALGO_DETERMINISTIC='1' such that we only use cuDNN deterministic algorithms.
> 
> I did not measure the performance of this solution but I'm getting deterministic results for gradient computation of convolutions, something that I was not getting previously
> 
> The code I used to test the determinism of the backward filter of the convolution is the following:
> 
> ```python
> import tensorflow as tf
> import numpy as np
> 
> np.random.seed(0)
> np.set_printoptions(precision=4)
> 
> def run_convolution_test(device, N=1):
>   tf.reset_default_graph()
>   tf.set_random_seed(42)
>   with tf.device(device):
>     input = tf.random_normal([1, 2048, 2048, 1], name='input')
>     kernel = tf.get_variable('kernel', shape = [3,3,1,1])
>     conv = tf.nn.conv2d(name='conv', input=input,
>                         filter=kernel, strides=[1,1,1,1], padding='SAME',
>                         use_cudnn_on_gpu=True)
>     grad = tf.gradients(conv, kernel)
>   grad_val = None
>   a_val = None
>   config = tf.ConfigProto(intra_op_parallelism_threads=1, inter_op_parallelism_threads=1)
>   with tf.Session(config=config) as sess:
>     sess.run(tf.global_variables_initializer())
>     conv_val, grad_val = sess.run([tf.squeeze(conv), tf.squeeze(grad)])
>     return conv_val, grad_val
> 
> def compute_error_matrix(matrices):
>     total_absolute_error = np.zeros((len(matrices), len(matrices)))
>     for i in range(len(matrices)):
>         for j in range(len(matrices)):
>             if i != j:
>                 dif=matrices[i] - matrices[j]
>                 total_absolute_error[i][j] = np.sum(np.absolute(dif))
>     print(total_absolute_error)
> 
> print("Running on GPU")
> conv_gpu, grad_gpu = zip(*[run_convolution_test('/device:GPU:0', i) for i in range(10)])
> compute_error_matrix(conv_gpu)
> compute_error_matrix(grad_gpu)
> 
> print("Running on CPU")
> conv_cpu, grad_cpu = zip(*[run_convolution_test('/device:CPU:0', i) for i in range(10,20)])
> compute_error_matrix(conv_cpu)
> compute_error_matrix(grad_cpu)
> ```

Unfortunately I cannot reproduce the deterministic computation you're claiming to achieve :(
For some reason, this FLAG is not supported (yet?).

I am using Python `3.6` and TensorFlow `1.11` compiled from source on `CUDA 10.0` + `CuDnn 7.2`.

Any clue?

**Comment by sidharths** [comment]
Is there a solution to this?

**Comment by UCRajkumar** [comment]
@chithangduong How are you able to send just `clip_by_value` to the CPU and keep everything else the same? I'm using Keras to do my training. Is there a way to still do it or do I need to switch into tf?

**Comment by NicolasSofu** [comment]
in march 2019 the lack of deterministic behaviour of the adam optimizer is still a problem !
 I don't want to force the optimizer to run on CPU (i don't want to wait for the rest of my life) ....

Is there any new proposal ?
Or is the "clip_by_value" on cpu a real answer ? if so, can someone detail it a little because i didn't catch this.

**Comment by NicolasSofu** [comment]
> We are closing this issue for now due to lack of activity. Please comment if this is still an issue for you. Thanks!

yes still an issue for me at least :+1:

**Comment by evolu8** [comment]
Still an issue!

**Comment by alexaorrico** [comment]
I'm an educator trying to teach students Tensorflow. This is a significant issue for me as the non-deterministic nature of Tensorflow on GPU makes it extremely difficult to automate the checking of students' projects. +1 @OutSorcerer, having a setting like in Theano to allow for deterministic behavior would be much appreciated.

**Comment by timespacce** [comment]
The non determinstic behaviour of cudnn is caused by the non deterministic algorithms that are being used. Those are sometimes faster - so the trade-off is between mathematical accuracy and training time. Those algorithms compute sometimes gradients that are leading to better model accuracy.

I agree - it would have been very nice if we had a feature to choose the algorithms we would like to use. Adding a setting in Conv2D that selects convolution algorithms and leading this one to the cuda_dnn.cc should not be that much of an effort.

Here is a workaround that works for me:
1. Download tensorflow source code
2. Go to the c++ class **cuda_dnn.cc**
3. Go to the function **ToConvBackwardFilterAlgo**
4. Make it return **cudnnConvolutionBwdFilterAlgo_t::CUDNN_CONVOLUTION_BWD_FILTER_ALGO_1;** or any other of the following algorithms that is deterministic and implemented:
    **CUDNN_CONVOLUTION_BWD_FILTER_ALGO_0**  (non-deterministic)
    **CUDNN_CONVOLUTION_BWD_FILTER_ALGO_1**
    **CUDNN_CONVOLUTION_BWD_FILTER_ALGO_FFT**
    **CUDNN_CONVOLUTION_BWD_FILTER_ALGO_3** (non-deterministic)
    **CUDNN_CONVOLUTION_BWD_FILTER_ALGO_WINOGRAD** (not implemented)
    **CUDNN_CONVOLUTION_BWD_FILTER_ALGO_WINOGRAD_NONFUSED**
    **CUDNN_CONVOLUTION_BWD_FILTER_ALGO_FFT_TILING**
    **CUDNN_CONVOLUTION_BWD_FILTER_ALGO_COUNT**
5. Build and compile the source code with bazel and install with pip.

Don't forget that this is only the computation of the gradients. In order to have an absolute determinstic behaviour all other parts of the model should be deterministic too.

**Comment by frankcoumans** [comment]
Check out the tf-deterministic package. From tensorflow 1.14 onwards it should mitigate quite a lot of GPU related non determinism.

**Comment by Sharon507** [comment]
Yes - this looks like it could be the solution, although I haven't tested it yet:
https://pypi.org/project/tensorflow-determinism/
